### PR TITLE
feat(manager/helm-values): Add support for bumpVersion

### LINKED ISSUE #4728: Helm dependencies should optionally bump chart.yaml
**What would you like Renovate to be able to do?**
It is great that renovate supports helm dependency management now. But, it would be great if renovate can optionally bump the chart version in chart.yaml in the PR as the new chart needs to be published as a new vesion.

**Describe the solution you'd like**
Renovate should (optionally ) bump the chart version in Chart.yaml when raising a PR to bump requirements.

**Describe alternatives you've considered**
Currently one way is to manually edit it so that we can publish a new version of chart.

**Additional context**
NA

### LINKED ISSUE #8231: Support for bumpVersion when updating helm-values
**What would you like Renovate to be able to do?**

When the container image tags in Helm value files are updated using the `helm-values` manager, it should also support bumping the chart version.

**Did you already have any implementation ideas?**

I've been attempting this in https://github.com/chgl/renovate/tree/add-bumpversion-support-to-helm-values (Let me know if you'd like a WIP PR). But the main challenge I face is that the `helm-values` manager works on `values.yaml` files, but the `packageFile` containing the version that should be bumped is `Chart.yaml`. I can work around this by opening and reading the `Chart.yaml` file and get as far as updating the file contents with a bumped version (https://github.com/chgl/renovate/blob/add-bumpversion-support-to-helm-values/lib/manager/helm-values/update.spec.ts#L38), but now I am not sure how to correctly write back the updated contents without majorly updating `getUpdatedPackageFiles` (it currently incorrectly overwrites `values.yaml` with the updated contents of `Chart.yaml`). 

I'm definitely interested in working on this.

### PR DESCRIPTION
**Pull request by kvanzuijlen**
## Changes
Adds support for `bumpVersion` to the helm-values manager.

## Context
Closes #4728, closes #8231

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/workers/repository/update/branch/get-updated.ts [hunk 1]
```diff
@@ -174,24 +186,39 @@ export async function getUpdatedPackageFiles(
...
+            if (bumpedPackageFiles) {
+              for (const bumpedPackageFile of bumpedPackageFiles) {
+                logger.debug(
+                  { bumpedPackageFile, depName },
+                  'Updating bumpedPackageFile content',
+                );
+                updatedFileContents[bumpedPackageFile.fileName] =
+                  bumpedPackageFile.newContent;
+              }
+            }
```
Comment (kvanzuijlen) [inline lib/workers/repository/update/branch/get-updated.ts:221]: This is the biggest difference; I included this part in the if-statement.

**Review by kvanzuijlen — COMMENTED** [review]
@rarkins I could reuse most parts from #8240. I was able to reproduce the behavior that caused it to be reverted.

Do you want to include helm-requirements in the same PR or should that (optionally) be added in a separate PR? I don't think it's really necessary, as helmv3 has long been the standard already and as far as I understand, helm-requirements is for Chart dependencies in helmv2 (correct?).

File: lib/modules/manager/helm-values/update.spec.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+      image: {
+        registry: 'docker.io',
+        repository: 'docker/whalesay',
+        tag: '1.0.0',
+      },
+    });
+
+    beforeEach(() => {
+      jest.resetAllMocks();
+      fs.readLocalFile = jest.fn();
```
Comment (viceice) [inline lib/modules/manager/helm-values/update.spec.ts:22]: wrong, please checkout other tests for how to mock fs

File: lib/modules/manager/helm-values/update.spec.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+          helmValuesContent,
+          '0.0.2',
+          'patch',
+          'values.yaml',
+        );
+      expect(bumpedContent).toEqual(helmValuesContent);
+      expect(bumpedFiles).toHaveLength(1);
+      const bumpedFile = bumpedFiles![0];
+      expect(bumpedFile.fileName).toBe('Chart.yaml');
+      expect(bumpedFile.newContent).toMatchSnapshot();
```
Comment (viceice) [inline lib/modules/manager/helm-values/update.spec.ts:38]: no snapshots

File: lib/modules/manager/helm-values/util.ts [hunk 4]
```diff
@@ -41,3 +44,51 @@ export function matchesHelmValuesInlineImage(
...
+export async function getParsedSiblingChartYaml(
+  fileName: string,
+): Promise<any> {
+  try {
+    const chartContents = await getSiblingChartYamlContent(fileName);
+    if (!chartContents) {
+      logger.debug({ fileName }, 'Failed to find helm Chart.yaml');
+      return null;
+    }
+    const chart = yaml.load(chartContents, { json: true }) as any;
```
Comment (viceice) [inline lib/modules/manager/helm-values/util.ts:81]: use our yaml utils and zod schema validation

**Comment by rarkins** [comment]
Different managers can be done separately

**Comment by kvanzuijlen** [comment]
Thanks for taking the time to review my PR @viceice. I hope my changes are correct. Let me know if I need to change anything else!

File: lib/modules/manager/helm-values/util.ts [hunk 5]
```diff
@@ -41,3 +44,51 @@ export function matchesHelmValuesInlineImage(
...
+export async function getParsedSiblingChartYaml(
+  fileName: string,
+): Promise<ChartDefinition | null> {
+  try {
+    const chartContents = await getSiblingChartYamlContent(fileName);
+    if (!chartContents) {
+      logger.debug({ fileName }, 'Failed to find helm Chart.yaml');
+      return null;
+    }
+    const chart = parseSingleYaml(chartContents) as ChartDefinition;
```
Comment (viceice) [inline lib/modules/manager/helm-values/util.ts:81]: please also use zod schema validation

File: lib/workers/repository/update/branch/get-updated.ts [hunk 6]
```diff
@@ -174,24 +186,39 @@ export async function getUpdatedPackageFiles(
...
           }
           if (res === packageFileContent) {
             logger.debug({ packageFile, depName }, 'No content changed');
           } else {
             logger.debug({ packageFile, depName }, 'Contents updated');
             updatedFileContents[packageFile] = res!;
             delete nonUpdatedFileContents[packageFile];
+            // indicates that the version was bumped in one or more files in
+            // addition to or instead of the packageFile
+            if (bumpedPackageFiles) {
```
[the remark this answers is not in the document]
Comment (secustor) [reply lib/workers/repository/update/branch/get-updated.ts:212]: I agree changes are here are risky and should be on a separate PR.

[the remark this answers is not in the document]
Comment (kvanzuijlen) [reply lib/workers/repository/update/branch/get-updated.ts:212]: Somehow I was under the impression that this piece of code was needed to make it all work. Tried running Renovate against a real repository though and found out that it's not needed at all. Going to revert the changes to get-updated.ts (and its tests) and zod parsing shortly.

File: lib/modules/manager/helm-values/update.spec.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    const helmValuesContent = yaml.dump({
+      image: {
+        registry: 'docker.io',
+        repository: 'docker/whalesay',
+        tag: '1.0.0',
+      },
+    });
+
+    beforeEach(() => {
+      jest.resetAllMocks();
```
Comment (viceice) [inline lib/modules/manager/helm-values/update.spec.ts:23]: I think this should be removed

File: lib/modules/manager/helm-values/types.ts [hunk 8]
```diff
@@ -18,3 +21,17 @@ export interface HelmDockerImageDependencyVersion
...
+
+export const ChartDefinition = z
+  .object({
+    apiVersion: z.string().regex(/v([12])/),
+    name: z.string().min(1),
+    version: z.string().regex(
+      // Copied from https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
+      /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$/,
+    ),
+  })
```
Comment (viceice) [inline lib/modules/manager/helm-values/types.ts:33]: remove the regex, we don't need to be so strict here. a wrong version will be catched on bump version.

File: lib/modules/manager/helm-values/types.ts [hunk 9]
```diff
@@ -18,3 +21,17 @@ export interface HelmDockerImageDependencyVersion
 export type HelmDockerImageDependency =
   | HelmDockerImageDependencyTag
   | HelmDockerImageDependencyVersion;
+
+export const ChartDefinition = z
```
Comment (viceice) [inline lib/modules/manager/helm-values/types.ts:25]: we prefer to have zod schema definitions in a schema.ts file, so please move

File: lib/workers/repository/update/branch/get-updated.ts [hunk 10]
```diff
@@ -20,6 +20,21 @@ export interface PackageFilesResult {
   updatedArtifacts: FileChange[];
 }
 
+async function getFileContent(
```
Comment (viceice) [inline lib/workers/repository/update/branch/get-updated.ts:23]: these changes should be done before in separate PR

**Comment by kvanzuijlen** [comment]
Thanks for all your patience @rarkins. I made the requested changes.

File: lib/modules/manager/helm-values/update.spec.ts [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+        '0.0.2',
+        true as any,
+        'values.yaml',
+      );
+      expect(bumpedContent).toEqual(helmValuesContent);
+    });
+
+    it('returns content if retrieving Chart.yaml fails', async () => {
+      jest.resetAllMocks();
+      fs.readLocalFile.mockRejectedValueOnce(null);
```
Comment (viceice) [inline lib/modules/manager/helm-values/update.spec.ts:68]: you just need to reset the read local file mock

File: lib/workers/repository/update/branch/get-updated.ts [hunk 12]
```diff
@@ -183,6 +183,7 @@ export async function getUpdatedPackageFiles(
               res,
               upgrade.packageFileVersion,
               upgrade.bumpVersion,
+              packageFile,
```
Comment (viceice) [inline lib/workers/repository/update/branch/get-updated.ts:186]: still a change which should be in a separate PR

File: lib/workers/repository/update/branch/get-updated.ts [hunk 13]
```diff
@@ -217,6 +218,7 @@ export async function getUpdatedPackageFiles(
           newContent,
           upgrade.packageFileVersion,
           upgrade.bumpVersion,
+          packageFile,
```
Comment (viceice) [inline lib/workers/repository/update/branch/get-updated.ts:221]: too

**Review by viceice — CHANGES_REQUESTED** [review]
blocked by refactor PR

**Review by viceice — APPROVED** [review]

**Pull request merged**